### PR TITLE
Frontend

### PR DESCRIPTION
**Pull request by helenmedrano**
Boilerplate for frontend setup
Includes routing (empty right now)
And config to proxy server/other needs
Includes linter

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    "@types/react-dom": "16.9.0",
+    "@types/react-router-dom": "^4.3.5",
+    "@types/styled-components": "^4.1.19",
+    "axios": "^0.19.0",
+    "ramda": "^0.26.1",
+    "react": "^16.9.0",
+    "react-dom": "^16.9.0",
+    "react-router-dom": "^5.0.1",
+    "react-scripts": "3.1.x",
+    "styled-components": "^4.3.2",
```
Comment (helenmedrano) [inline frontend/package.json]: could also just use `.css` files?

Comment (helenmedrano) [reply frontend/package.json]: But I like for readability sometimes

Comment (wmedrano) [reply frontend/package.json]: It seems LEGO uses styled-components so 🍍

File: frontend/src/App.test.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import App from './App';
+
+it('renders without crashing', () => {
```
Comment (wmedrano) [inline frontend/src/App.test.js]: Where does this come from? Is there documentation for the library where `it` comes from.

File: frontend/package.json (hunk 1 above)
Comment (wmedrano) [reply frontend/package.json]: If it doesn't work out we can refactor

File: frontend/src/models/plugin.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+export type DefaultParamsType = {
+  volume: number,
+}
+
+export type PluginShape = {
+  uri: string,
+  name: string,
+  pluginType: string,
+  description: string,
```
Comment (wmedrano) [inline frontend/src/models/plugin.ts]: `description` before `pluginType` to be more consistent with `api.yaml`?

Or `abc` after `uri`.

File: frontend/src/models/plugin.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  defaultParams: DefaultParamsType
```
Comment (wmedrano) [inline frontend/src/models/plugin.ts]: This can be any arbitrary `JSON`. Can it be type `[key: string]: any`? We will need to add some extra metadata per plugin to let the frontend know the shape when configuring.

Also, seems like `api.yaml` is out of date. The rust schema has defaultParams: https://github.com/wmedrano/wdaw/blob/master/api/rust/src/schemas.rs#L63

Can you update `api.yaml` to have defaultParams be of type `Object`?
https://github.com/wmedrano/wdaw/blob/master/api/api.yaml#L267

File: frontend/src/services/api_request.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
+import * as R from 'ramda'
+import axios from 'axios'
+import { PathNotFoundErr, ParamRequiredErr } from './api_request_error'
+
+// TODO:
+// Merge all the type to go-api lib
```
Comment (wmedrano) [inline frontend/src/services/api_request.ts]: What is `go-api` lib?

**Comment by wmedrano** [comment]
The current implementation makes /plugin_manager a whole route. Will we still be able to easily embed the component into another view.

Example:
image
Ableton has a sidebar of resources. You can select the plugins category and a list of plugins will appear.

File: frontend/src/models/plugin.ts (hunk 4 above)
Comment (wmedrano) [reply frontend/src/models/plugin.ts]: OpenAPI uses [JSON Schema](https://json-schema.org/understanding-json-schema/reference/object.html) to describe types. Here is documentation about the `Object` type.

TLDR: I think it will end up being:
```yaml
Params:
  type: Object
  additionalProperties: { "type": "string" }
```

File: frontend/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
Comment (wmedrano) [inline frontend/tsconfig.json]: seems like es10 is out.

Are there any advantages or disadvantages to using the newer ones?

File: frontend/src/services/plugin_service/index.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+export default class PluginService {
+  apiRequest: ApiRequest
+
+  constructor() {
+    this.apiRequest = new ApiRequest({ pathMap, baseURL: '' })
+  }
+
+  getPlugins(): Promise<Array<Plugin>> {
+    return this.apiRequest.get('plugins').then(({ data }: GetPluginsResponse) => {
```
Comment (wmedrano) [inline frontend/src/services/plugin_service/index.ts]: Should this not be this.apiRequest.plugins.then ...?

I think it would be better if it gives you the compile time checks.

File: frontend/src/pages/plugin_manager/types.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import Plugin from '../../models/plugin'
+
+export type StateType = {
+  plugins: Array<Plugin>
```
Comment (wmedrano) [inline frontend/src/pages/plugin_manager/types.ts]: Is this used? Looks like PluginManager gets the set of plugins through an HTTP request.

File: frontend/src/services/api_request.ts (hunk 5 above)
Comment (helenmedrano) [reply frontend/src/services/api_request.ts]: accidenty

File: frontend/src/models/plugin.ts (hunk 4 above)
Comment (helenmedrano) [reply frontend/src/models/plugin.ts]: will leave for later PR

File: frontend/src/App.test.js (hunk 2 above)
Comment (helenmedrano) [reply frontend/src/App.test.js]: it is just exist. Testing can be augmented with mocha/chai/enzyme and whatever else

File: frontend/tsconfig.json (hunk 6 above)
Comment (helenmedrano) [reply frontend/tsconfig.json]: nah it's just what create react app did
